### PR TITLE
add newrelic integration

### PR DESCRIPTION
**Pull request by itajaja**

### CODE REVIEW

**Comment by itajaja** [comment]
This will both report Transactions in APM and custom metrics in Insights

**Comment by itajaja** [comment]
but still in this repo, right? I'll have to move the code around a little (add some hooks here and there) but I think I can

**Comment by itajaja** [comment]
like that?

File: tqp/newrelic.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+# -----------------------------------------------------------------------------
+
+
+def install():
+    QueuePollerBase._handle_message = newrelic.agent.background_task()(
+        QueuePollerBase._handle_message,
+    )
+
+    super_handle_message = TopicQueuePoller.handle_message
```
Comment (taion) [inline tqp/newrelic.py]: i would call this `base_handle_message`

File: tqp/newrelic.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    super_handle_error = TopicQueuePoller.handle_error
```
Comment (taion) [inline tqp/newrelic.py]: `base_handle_error`

File: tqp/newrelic.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+    def _hook_handle_message(self, raw_msg, payload):
```
Comment (taion) [inline tqp/newrelic.py]: probably name these just as the names of the overridden methods

you may also want to use `@functools.wraps`

File: tqp/topic_queue_poller.py [hunk 4]
```diff
@@ -116,28 +120,38 @@ def __init__(self, *args, **kwargs):
...
+        meta = payload['meta']
+        message = payload['message']
+
+        self.logger.info('%s: handling new message', topic)
+        self.logger.debug(raw_msg.body)
+
+        extra_call_kwargs = {'meta': meta} if meta is not None else {}
+        handler(message, **extra_call_kwargs)
+
+    def handle_error(self, ex, raw_msg, payload):
```
Comment (taion) [inline tqp/topic_queue_poller.py]: needed?

File: tqp/topic_queue_poller.py [hunk 5]
```diff
@@ -65,23 +65,27 @@ def extend_message_visibility_timeout(self, queue, messages, timeout):
     def ensure_queue(self):
         return create_queue(self.queue_name, **self.queue_attributes)
 
-    def _handle_message(self, msg):
+    def parse_raw_message(msg):
```
Comment (taion) [inline tqp/topic_queue_poller.py]: `get_message_payload`?

File: tqp/topic_queue_poller.py [hunk 6]
```diff
@@ -65,23 +65,27 @@ def extend_message_visibility_timeout(self, queue, messages, timeout):
...
+        return None
+
+    def _handle_message(self, raw_msg):
```
Comment (taion) [inline tqp/topic_queue_poller.py]: then you can keep this named `message` or `msg`

File: tqp/topic_queue_poller.py [hunk 7]
```diff
@@ -65,23 +65,27 @@ def extend_message_visibility_timeout(self, queue, messages, timeout):
...
+        payload = self.parse_raw_message(raw_msg)
         try:
-            self.handle_message(msg)
+            self.handle_message(raw_msg, payload)
 
-            msg.delete()
+            raw_msg.delete()
             self.logger.debug('message successfully deleted')
-        except Exception:
+        except Exception as ex:
```
Comment (taion) [inline tqp/topic_queue_poller.py]: what's wrong with calling this `e` like everywhere else 😛

File: tqp/topic_queue_poller.py [hunk 8]
```diff
@@ -65,23 +65,27 @@ def extend_message_visibility_timeout(self, queue, messages, timeout):
...
             # whatever the error is, log and move on
-            self.handle_error(msg)
+            self.handle_error(ex, raw_msg, payload)
 
-    def handle_error(self, msg):
+    def handle_error(self, ex, raw_msg, payload):
```
Comment (taion) [inline tqp/topic_queue_poller.py]: `msg, payload`?

**Pull request merged**